File: 0078-subsets/0078-subsets.js
var subsets = function (nums) {
	let output = [[]];
	for (let num of nums) {
		let newSubsets = [];
		for (let curr of output) {
			let temp = curr.slice();
			temp.push(num);
			newSubsets.push(temp);
		}
		for (let curr of newSubsets) {
			output.push(curr);
		}
	}
	return output;
};

File: Python/0078-subsets.py
# time complexity: O(n*2^n)
# space complexity: O(n*2^n)
from itertools import chain, combinations
from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        result = list(chain.from_iterable(combinations(nums, r)
                      for r in range(len(nums)+1)))
        for i, item in enumerate(result):
            result[i] = list(item)
        return result

# class Solution:
#    def subsets(self, nums: List[int]) -> List[List[int]]:
#        output = [[]]

#        for num in nums:
#            output += [curr + [num] for curr in output]

#        return output


nums = [1, 2, 3]

print(Solution().subsets(nums))
